nri-postgresql/src/Postgres/Enum.hs

```diff
@@ -0,0 +1,242 @@
+              )
+
+      _ ->
+        Prelude.fail <| "The datatype name " ++ quote hsTypeString ++ " must be a data declaration."
+
+  -- Check for duplicate values specified in the mapping
+  _ <- 
+    case findDuplicates (List.map Tuple.second mapping) of 
+      [] -> 
+        Prelude.pure () 
+
+      duplicates ->
+        Prelude.fail <| "The following values in the mapping are duplicated: " ++ Text.toList (makeValueList duplicates)  
+
+  -- Check for duplicate constructor names specified in the mapping
+  _ <-
+    case findDuplicates (List.map Tuple.first mapping) of
+      [] -> 
+        Prelude.pure ()
+
+      duplicates ->
+        Prelude.fail <| "The following constructors in the mapping are duplicated: " ++ Text.toList (makeConList duplicates)
+
+  let hsConSet = Set.fromList conNames
+  let mappingConSet = Set.fromList (List.map Tuple.first mapping)
+
+  _ <- 
+    case (Set.toList (Set.diff hsConSet mappingConSet), Set.toList (Set.diff mappingConSet hsConSet)) of 
+      ([], []) -> 
+        Prelude.pure ()
+
+      ([], mappingOnlyCons) ->
+        Prelude.fail <| "The following names in the mapping are not constructors from the data type " ++ quote hsTypeString ++ ": " ++ Text.toList (makeConList mappingOnlyCons)
+
+      (hsOnlyCons, _) -> 
+        Prelude.fail <| "The following constructors are not present in the mapping: " ++ Text.toList (makeConList hsOnlyCons)
+
+  pgDatabase <-
+    TH.runIO <| do 
+      nriSettings <- Environment.decode Postgres.Settings.decoder
+      Prelude.pure (Postgres.Settings.toPGDatabase nriSettings)
+
+  dbConnectDecls <- useTPGDatabase pgDatabase
+
+  (pgEnumValues :: List Text) <-
+    TH.runIO <| withTPGConnection (\connection -> do 
+      -- Check if the databaseTypeName exists on the PG database and is an enum
+      -- See https://www.postgresql.org/docs/current/catalog-pg-type.html
+      (_, typeRows) <-
+        pgSimpleQuery connection (BSL.fromChunks
+          [ "SELECT typtype"
+          , " FROM pg_catalog.pg_type"
+          , " WHERE typname = '", Encoding.encodeUtf8 databaseTypeName, "'"
+          ])
+
+      _ <- case List.map (\[v] -> pgDecodeRep v) typeRows of 
+        [] -> 
+          Prelude.fail ("Type " ++ quote (Text.toList databaseTypeName) ++ " does not exist on the database.")
+        
+        -- 'e' means enum type
+        ['e'] -> 
+          Prelude.pure ()
+        
+        _ -> 
+          Prelude.fail ("Type " ++ quote (Text.toList databaseTypeName) ++ " is not an enum type.")
+
+      (_, enumRows) <- 
+        pgSimpleQuery connection (BSL.fromChunks
+          [ "SELECT enumlabel"
+          , " FROM pg_catalog.pg_enum"
+          , " WHERE enumtypid = '", Encoding.encodeUtf8 databaseTypeName, "'::regtype"
+          , " ORDER BY enumsortorder"
+          ])
+        
+      case List.map (\[value] -> pgDecodeRep value) enumRows of 
```

Comment:
wanna filter map and case on the list here?

Reply:
Sure, was following the lead of the library - I guess the filterMap isn't that much uglier.